src/metabase/query_processor/streaming/xlsx.clj

```diff
@@ -392,25 +393,58 @@
         (str column-title " (" (currency-identifier col-viz-settings) ")")
         column-title))))
 
+(def ^:private auto-sizing-threshold
+  "The number of rows used for auto-sizing columns. If this number is too large, exports of large datasets
+  will be prohibitively slow."
+  1000)
+
+(def ^:private extra-column-width
+  "The extra width applied to columns after they have been auto-sized, in units of 1/256 of a character width.
+  This ensures the cells in the header row have enough room for the filter dropdown icon."
+  (* 4 256))
+
+(defn- autosize-columns!
+  "Adjusts each column to fit its largest value, plus a constant amount of extra padding."
+  [sheet col-count]
+  (doseq [i (range col-count)]
+    (.autoSizeColumn ^SXSSFSheet sheet i)
+    (.setColumnWidth ^SXSSFSheet sheet i (+ (.getColumnWidth ^SXSSFSheet sheet i) extra-column-width)))
+  (.untrackAllColumnsForAutoSizing ^SXSSFSheet sheet))
+
+(defn- setup-header-row!
+  "Turns on auto-filter for the header row, which adds a button to each header cell that allows columns to be
+  filtered and sorted. Also freezes the header row so that it floats above the data."
+  [sheet col-count]
+  (when (> col-count 0)
+    (.setAutoFilter ^SXSSFSheet sheet (new CellRangeAddress 0 0 0 (dec col-count)))
+    (.createFreezePane ^SXSSFSheet sheet 0 1)))
+
 (defmethod i/streaming-results-writer :xlsx
   [_ ^OutputStream os]
   (let [workbook            (SXSSFWorkbook.)
         sheet               (spreadsheet/add-sheet! workbook (tru "Query result"))]
     (reify i/StreamingResultsWriter
       (begin! [_ {{:keys [ordered-cols]} :data} {col-settings ::mb.viz/column-settings}]
+        (.trackAllColumnsForAutoSizing ^SXSSFSheet sheet)
+        (setup-header-row! sheet (count ordered-cols))
         (spreadsheet/add-row! sheet (column-titles ordered-cols col-settings)))
 
-      (write-row! [_ row _ ordered-cols {:keys [output-order] :as viz-settings}]
+      (write-row! [_ row row-count ordered-cols {:keys [output-order] :as viz-settings}]
         (let [ordered-row  (if output-order
                              (let [row-v (into [] row)]
                                (for [i output-order] (row-v i)))
                              row)
               col-settings (::mb.viz/column-settings viz-settings)
               cell-styles  (cell-style-delays workbook ordered-cols col-settings)]
           (binding [*cell-styles* cell-styles]
-            (add-row! sheet ordered-row ordered-cols col-settings))))
-
-      (finish! [_ _]
+            (add-row! sheet ordered-row ordered-cols col-settings))
+          (when (= row-count auto-sizing-threshold)
+            (autosize-columns! sheet (count row)))))
```

Comment:
The more I think about it, the more I think `row-count` isn't really accurate at all since `row-count` for the first row is `0`.

So in this case `write-row!` is calling `autosize-columns!` on the 1001th row... this should be calling it on the 1000th row, right?

If that was fixed tho, since `finish!` _also_ calls `autosize-columns!` if the number of rows is <= 1000, you'd end up with 2 calls to `autosize-columns!` if the number of rows was exactly 1000. Since `autosize-columns!` calls `.untrackAllColumnsForAutoSizing`, would the subsequent call to `autosize-columns!` fail, because the columns are no longer tracked?

If there's a `getTrackAllColumnsForAutoSizing` method or something like that maybe you should have `autosize-columns!` check that so it no-ops if it can't actually autosize stuff.

Reply:
Hmm, yeah good point. There's a `getTrackedColumnsForAutoSizing`; I'll see if that'll work.